engine.py: attempt to remove unnecessary files

glacambre opened the pull request:
This commit makes adareducer attempt to randomly delete as many files as
possible before and after doing reduction.

Closes #17.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/AdaCore/adareducer?pullRequest=20) <br/>All committers have signed the CLA.

setton commented on adareducer/engine.py:
```diff
@@ -82,6 +91,22 @@ def run_predicate(self, print_if_error=False):
             log(out.stdout.decode() + "\n" + out.stderr.decode())
         return status
 
+    def attempt_delete_all(self, files):
```
needs doc :)

setton commented on adareducer/engine.py:
```diff
@@ -194,14 +219,31 @@ def next_file_to_process(self):
     def run(self):
         """Run self: reduce the project as much as possible"""
 
+        all_files = set([os.path.abspath(f) for f in os.listdir(self.project_dir)])
```
Now that we have Proper project support, let's do something different to get the full list: we can either only focus on trying to delete all project source files, or we could go one step further (as you're doing right now) and look at all files in all source dirs of the project tree. As you wish.

The object self.resolver is a ProjectResolver (see project_support.py): this object should be augmented to give us the list of source dirs if we want to go that route.

setton replied:
Note that I have a slight preference for attempting project sources only, this saves us from having to be careful about predicate artifacts or adareducer artifacts, while still being useful for the majority of use case.

glacambre replied:
From a user perspective I think that I would prefer to have something that cleans up as many files as possible, leaving me with something that I could directly file in my testsuite.

But I now realize that this is in fact very dangerous and that this could remove files in, e.g. `.git`, which is certainly something we don't want. So I'll follow your advice.

setton commented on adareducer/engine.py:
```diff
@@ -194,14 +219,31 @@ def next_file_to_process(self):
+        if DELETE_FILES_BEFORE:
+            to_delete = [
+                f
+                for f in (all_files - predicate_artifacts - self.nonremovable_files)
```
here we would need to protect from deletion any ".orig" file and any ".crash" file that could be generated by previous runs of adareducer - this is in the case that adareducer crashes and the user wants to salvage the run and fix manually.

setton commented on adareducer/engine.py:
```diff
@@ -228,6 +270,21 @@ def run(self):
+        # Try to brute-delete files again. This might give us new results when
+        # compared to the first time we called attempt_delete_all thanks to the
+        # project being reduced.
+        if DELETE_FILES_AFTER:
```
I think this step can be removed completely: as part of self.reduce_file we're trying to delete each source already

setton commented on testsuite/tests/deletion/test.yaml:
```diff
@@ -0,0 +1 @@
+description: "minimal"
```
the description should be changed to "deletion" (I wish e3-testsuite didn't force us to write these yaml files but that's anoter project for another day)

glacambre commented:
I'm also wondering, would it make sense to keep the files renamed ".deleted" and to just let the user do something like `find . -iname "*.deleted" -delete`? This would keep in line with adareducer keeping `.orig` files.

setton commented:
> I'm also wondering, would it make sense to keep the files renamed ".deleted" and to just let the user do something like `find . -iname "*.deleted" -delete`? This would keep in line with adareducer keeping `.orig` files.

Yes it makes sense to keep them. The idea being ".orig" files is to allow recovery in case adareducer crashes. When adareducer is a bit more battle-tested, we can add a pass to remove these at the end of a successful run, at which point we'll remove the "deleted" ones too.

glacambre commented:
Done :)
